coordinator batching

tbg opened the pull request:
Glued everything together, rebased and fixed bugs.
All tests pass, but there is still some stabilizing to do (`kv` takes long/timeouts sometimes, likely an intent/push issue).

The big base commit is #2141 after rebase and resulting fixes, so it's mostly been "reviewed". Everything else is new.

Still some work to do, but mostly cleanup. @spencerkimball you probably want to take a look already.

tbg committed:
batch DistSender.Send to Store.ExecuteCmd

see #2130, #1998 for context.

* change `(*DistSender).Send` to wrap all requests in a single-
  element Batch (TxnCoordSender unwrapped any `Batch` previously).
* `(*Store).ExecuteCmd` unwraps this single-element `Batch`.
* verifyPermissions: check only the inside of BatchRequest.
* introduce `Reverse` flag on `BatchRequest` to use instead of checking for
  `ReverseScanRequest`. Goal: whole `Batch` is either `Reverse` or not, and
  any range-op in it is simply going to be using the reverse order. Not being
  able to mix reverse/non-reverse requests stems from the change in descriptor
  lookup logic.
* implemented `proto.Combinable` for `BatchRequest`; changed it to return an
`error`.
* implement logic to deal with `proto.Countable` and `proto.Bounded` in `Send`.
* batch expansion at (*Store.ExecuteCmd)
  updated ExecuteCmd so that it properly unrolls a Batch.
  Note that it has to do hacky things with `Key` and `EndKey`
  to make sure they are truncated appropriately if the `BatchRequest`
  is (after all, that's all `DistSender` adapts for cross-range ops).
  Other than that, we mostly propagate the whole header from the Batch
  and the Txn (if any) from each response to the next request.

  `TxnCoordSender` still unrolls Batches and `(*DistSender).Send` still
  wraps each individual request in a new Batch.

  Some store tests use batches "properly", and so does `resolveWriteIntentError`,
  so the new functionality is actually tested and it took a bit to get all
  tests to pass again.

  Note also that the first simplification was made: the test sender in ./storage
  lost its fake batch handling.
* complete Is{Read,Write,...} for BatchRequest
* add batch-compatible abstractions
  checking for EndTransaction and intents uses a (likely temp)
  helper that produces correct results for `BatchRequest`.
* proper truncation of batch at DistSender
  this change removes the provisional truncation code in
  `(*Store).ExecuteCmd` in favor of cleaned up code in
  `(*DistSender).Send/sendAttempt`, which has also seen
  some refactoring and many new TODOs.
* workaround for non-atomicity of batch on Store
  a retry half-way through a batch may leave an earlier CPut
  fail upon retry because it reads the intent it previously
  wrote.
* refactor range descriptor cache
  keep the recursion in LookupRange and adapt the
* retryableLocalSender -> DistSender
  remove retryableLocalSender and plug in a
  DistSender which uses a LocalSender as its
  rangeDescriptorDB.

  KV tests still fail occasionally, but this
  is likely due to lack of non-atomicity on
  Store.
  Now after rebase, they fail but I'm sure
  I can fix that.
  ./kv and ./storage pass with a few tests disabled (which are expected to fail)
  ./sql passes in 17s, have to check why
* {client->batch}.Sender
* refactor DistSender and TxnCoordSender
* send only batches from client
* don't send replies on errors (at least in {TxnCoord,Dist}Sender)
  we're now pretty close to req->(resp,err) everywhere.
* re-enable most storage tests
  changed multiTestContext and some other test contexts used
  to use a DistSender attached to a LocalSender (instead of
  previously only LocalSender). This allows re-enabling all
  of the tests which had to be disabled because they would
  involve batches which required multi-range logic not
  available without a DistSender inbetween.
* fix server.TestIntentResolution
* expand kv.TestTruncate, refactor truncate() slightly
* re-add auto txn wrap for cross-range reqs
* CmdID generation in ChunkingSender
* various test fixes
* fix RangeLookup when in reverse mode looking up EndKey of range with intent
* tweak RangeLookup logic so that empty results are never returned. Panic if it
  does happen.
* client workaround for request errors
  the client code wants to know which request failed, but
  with the (response, error) pattern no reply is returned
  on error, so no reply header can be used to find out the
  location of the error.
  this needs further fixing, see #1891.
* deactivate some assertions pending #2300
* disabled TestTree
* feedback and shape-up for merge
  attempt to remove issues slowing down tests.
  those likely arose since sending of batches
  is more sensitive to intent errors (since
  a single "command" now covers more keys)
* descNext -> bool in DistSender
* fixed various comments/TODOs
* fix rds<->kvs correspondence in RangeLookup when
  scanning extra descriptor in Reverse mode
* disable txn auto gc'ing (b/c store atomicity missing)
* make trace.Finalize() panic-aware (prints active trace to stdout, re-panics).
* carry an error over in the temporary batch unwrapping logic
* clarify the ConditionalPut hack
* disable two tests temporarily

tbg committed:
skip over irrelevant ranges in multi-range scan

tbg committed:
implement batch splitting

and remove the provisional logic in store.ExecuteCmd which
unrolled batches.

misc renaming

tbg committed:
fix timestamp updates to txn

which in turn fixes TestSingleKey (though it's a bit scary that's
the only test that caught it). The txn timestamp wasn't properly
updated from the response timestamp (so that reads/writes would
be pushed, but the Txn not updated, hence committing successfully).

tbg committed:
fix a bug in RangeLookup

when returning descriptors from intents, it was not
properly checked whether the intent had been obtained
through a forward scan during a reverse lookup, in which
case some scenarios demand that it be discarded.

removed spurious uses of KeyAddress and refactored some
of DistSender for better legibility.

tbg committed:
ignoreIntents -> considerIntents

tbg committed:
misc comment fixes and re-enabling of tests

tbg committed:
take package `batch` apart

moved some parts of it to kv. KeyRange is now keys.Range,
Short() is now String() on *BatchRequest in proto (changed so
that it doesn't import 'keys' any more), the rest mostly went
to 'client'.
Did a bunch of gorename (to unexport a few things). No change
of semantics in this commit.

tbg committed:
re-enable txn auto gc

tbg committed:
use (req)->(*resp, err) instead of (*req)->(*resp, err)

it is not supposed to be mutable, this is a first step towards that
(of course some data contained within is still shared, such as
requests and the Txn).

bdarnell commented on client/db.go:
```diff
@@ -451,27 +451,30 @@ func (db *DB) send(calls ...proto.Call) (err error) {
-		} else if c.Post != nil {
-			err = c.Post()
+			resetClientCmdID(c.Args)
+			db.sender.Send(context.TODO(), c)
```
Why does this code path call `db.sender.Send()` (and `resetClientCmdID`) instead of `db.send()` as on line 477?

tschottdorf replied:
this one is the real send call. The other one is calling itself after having wrapped in a batch.

tschottdorf commented:
This is starting to look good. I've worked more towards code cleanup and filling in some gaps. I want to completely understand why (mostly) the `kv` tests take longer (almost certainly has to do with pushes and intents) to make sure there's nothing sketchy going on.

Please review. It's unfortunately a huge change, sorry about that. The big base commit has been "observed" by a few of you and everything on top should be ok to review.

tbg committed:
un-hackify bounded request handling

requests which are saturated are now masked out via
NoopRequest, stopping once all requests are or no
more ranges are left to query.

tbg committed:
temporarily alter another logic test

bdarnell commented on client/txn_test.go:
```diff
@@ -118,22 +131,21 @@ func TestTxnRequestTxnTimestamp(t *testing.T) {
 	for testIdx = range testCases {
-		txn.db.sender.Send(context.Background(),
-			proto.Call{Args: testPutReq, Reply: &proto.PutResponse{}})
+		_ = sendCall(txn.db.sender, proto.Call{Args: testPutReq, Reply: &proto.PutResponse{}})
```
Why does this ignore the returned error?

tschottdorf replied:
added a check.

bdarnell commented on kv/batch.go:
```diff
@@ -0,0 +1,242 @@
+	var rplChunks []*proto.BatchResponse
+	var part []proto.RequestUnion
+	for len(parts) > 0 {
+		part, parts = parts[0], parts[1:]
```
Why isn't this just `for _, part := range parts`?

tschottdorf replied:
No reason. Changed.

On Mon, Sep 14, 2015 at 6:10 PM, Ben Darnell <notifications@github.com>
wrote:

> In kv/batch.go
> <https://github.com/cockroachdb/cockroach/pull/2406#discussion_r39453872>:
>
> > +		}
> > +	} else {
> > +		rng := rand.New(rand.NewSource(ba.CmdID.Random))
> > +		id := ba.CmdID
> > +		nextID = func() proto.ClientCmdID {
> > +			curID := id             // copy
> > +			id.Random = rng.Int63() // adjust for next call
> > +			return curID
> > +		}
> > +	}
> > +
> > +	parts := ba.Split()
> > +	var rplChunks []*proto.BatchResponse
> > +	var part []proto.RequestUnion
> > +	for len(parts) > 0 {
> > +		part, parts = parts[0], parts[1:]
>
> Why isn't this just for _, part := range parts?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/cockroachdb/cockroach/pull/2406/files#r39453872>.
>

bdarnell commented on kv/batch.go:
```diff
@@ -0,0 +1,242 @@
+}
+
+// SendBatch implements Sender.
+// TODO(tschottdorf): We actually don't want to chop EndTransaction off for
```
What is this comment referring to? There's nothing related to EndTransaction here now.

tschottdorf replied:
Changed to:

>
// TODO(tschottdorf): We actually don't want to chop EndTransaction off for
// single-range requests (but that happens now since EndTransaction has the
// isAlone flag). Whether it is one or not is unknown right now (you can
only
// find out after you've sent to the Range/looked up a descriptor that
suggests
// that you're multi-range. In those cases, the wrapped sender should
return an
// error so that we split and retry once the chunk which contains
// EndTransaction (i.e. the last one).


On Mon, Sep 14, 2015 at 6:12 PM, Ben Darnell <notifications@github.com>
wrote:

> In kv/batch.go
> <https://github.com/cockroachdb/cockroach/pull/2406#discussion_r39454068>:
>
> > +	return f(ctx, ba)
> > +}
> > +
> > +// A ChunkingSender sends batches, subdividing them appropriately.
> > +type chunkingSender struct {
> > +	f senderFn
> > +}
> > +
> > +// NewChunkingSender returns a new chunking sender which sends through the supplied
> > +// SenderFn.
> > +func newChunkingSender(f senderFn) client.BatchSender {
> > +	return &chunkingSender{f: f}
> > +}
> > +
> > +// SendBatch implements Sender.
> > +// TODO(tschottdorf): We actually don't want to chop EndTransaction off for
>
> What is this comment referring to? There's nothing related to
> EndTransaction here now.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/cockroachdb/cockroach/pull/2406/files#r39454068>.
>

bdarnell commented on storage/replica_command.go:
```diff
@@ -608,13 +633,15 @@ func (r *Replica) RangeLookup(batch engine.Engine, args proto.RangeLookupRequest
 
-	if args.IgnoreIntents && len(intents) > 0 {
+	var rds []proto.RangeDescriptor // corresponding unmarshaled descriptors
+	// TODO(tschottdorf): IgnoreIntents is only set on a Push, but there are
```
s/IgnoreIntents/ConsiderIntents/g

tschottdorf replied:
done.

On Mon, Sep 14, 2015 at 6:34 PM, Ben Darnell <notifications@github.com>
wrote:

> In storage/replica_command.go
> <https://github.com/cockroachdb/cockroach/pull/2406#discussion_r39455910>:
>
> >  	}
> >
> > -	if args.IgnoreIntents && len(intents) > 0 {
> > +	var rds []proto.RangeDescriptor // corresponding unmarshaled descriptors
> > +	// TODO(tschottdorf): IgnoreIntents is only set on a Push, but there are
>
> s/IgnoreIntents/ConsiderIntents/g
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/cockroachdb/cockroach/pull/2406/files#r39455910>.
>

bdarnell commented on kv/local_test_cluster.go:
```diff
@@ -146,5 +122,8 @@ func (ltc *LocalTestCluster) Start(t util.Tester) {
 
 // Stop stops the cluster.
 func (ltc *LocalTestCluster) Stop() {
+	if r := recover(); r != nil {
```
FYI this doesn't work very well. `t.Fatal()` calls `runtime.Goexit`, which is equivalent to `panic(nil)`, which you cannot detect with `recover()`. So it only partially solves the "hang at shutdown" problem. If we had the `testing.T` available we could do `if t.Failed() { return }`.

tschottdorf replied:
thanks for the heads up. It did solve some crashes for me - should I leave it in?

bdarnell replied:
I've tried it after seeing it in your PRs, and had mixed results with it. Sometimes it helps, by showing the original panic instead of a timeout, and sometimes it hurts, by losing the original panic's stack trace. I don't think it's worth adding this to all our stop methods; when I'm debugging something that might be affected by this I just comment out the `defer Stop()` line.

tschottdorf replied:
Ok, I'm going to take them out. For debugging spurious errors (where we currently have to search the goroutine stacks) it seems like we should disencourage using `Fatal{,f}` (or make really clear how to debug timeouts during tests).

bdarnell replied:
I strongly prefer `Fatalf()` to `Errorf()` as a general rule. If you have a panic after an `Errorf`, the `Errorf` message is thrown away and you only see the panic. We should only use `Errorf` when we are confident that continuing past an error will not result in a panic.

Where possible, we should add a `*testing.T` to our test structures that have a stop method. Then we can do this, which I think safely avoids the shutdown hangs for both panic and `Fatalf`:

    if t.Failed() {
        return
    }
    if r := recover(); r != nil {
        panic(r)
    }

tschottdorf replied:
I did that where I could (`grep -F 'panic(r)'`).

bdarnell commented:
LGTM. I didn't review it too thoroughly on this pass, but let's get it checked in (once that circleci failure is resolved) so we can move forward with the TODOs in new PRs.

tbg committed:
address @bdarnell's feedback

tbg committed:
move intent resolution **after** batch commit

this is relevant especially when resolving on EndTransaction:
only resolve **after** the new txn entry is persisted.

tbg committed:
save an unnecessary Push on external intents

tbg committed:
simplify RangeLookup logic

previously, we had special handling for PushRequest when it was in service of a
range lookup. This commit removes this logic in favor of the following: We
ignore intents but when encountering an addressing-related restart, ask for
them to be returned randomly.

tbg committed:
fix TestFailedReplicaChange

tschottdorf commented:
Ok. I'm worried about breaking other's builds (there have been more spurious test failures than I'm comfortable with). I added some commits which hopefully deal with it. Will rebuild this a bunch of times and if it's reliable we should be ok. Test times in `kv` (and in other places) are still way higher though:

>
TestEvictCacheOnError gateway_batch 0.55s
TestEvictCacheOnError master 0.55s
TestKVDBInternalMethods master 1.09s
TestRangeSplitsWithConcurrentTxns gateway_batch 0.20s
TestRangeSplitsWithConcurrentTxns master 0.14s
TestRangeSplitsWithWritePressure master 0.17s
TestRetryOnDescriptorLookupError gateway_batch 0.25s
TestRetryOnDescriptorLookupError master 0.25s
TestRetryOnWrongReplicaError gateway_batch 0.24s
TestTxnDBInconsistentAnalysisAnomaly gateway_batch 1.41s
TestTxnDBInconsistentAnalysisAnomaly master 0.92s
TestTxnDBLostUpdateAnomaly gateway_batch 0.65s
TestTxnDBLostUpdateAnomaly master 0.41s
TestTxnDBPhantomDeleteAnomaly gateway_batch 0.58s
TestTxnDBPhantomDeleteAnomaly master 0.41s
TestTxnDBPhantomReadAnomaly gateway_batch 0.75s
TestTxnDBPhantomReadAnomaly master 0.42s
TestTxnDBUncertainty gateway_batch 0.18s
TestTxnDBWriteSkewAnomaly gateway_batch 1.84s
TestTxnDBWriteSkewAnomaly master 1.14s

It's not clear to me whether this is expected but it gets harder to investigate once we land this (the anomaly tests are unfortunately not fun to debug, and prolonged test durations I've eliminated so far were due to push/intent issues).

tbg committed:
avoid some more hangs when tests Fatal{,f}

see https://github.com/cockroachdb/cockroach/pull/2406#discussion_r39550086

tschottdorf commented:
Ok, doesn't seem like the tests actually take longer, at least I'm seeing `kv` and others take about as long on `master`. I've built ~8x and got a green build on all of them (there was one obscure data race once, but it was related to debug logging and was missing a stack trace, so it was hard to debug - looked like printing an error that was also active in some raft goroutine, so not really tied to the change). So I'm going to merge this in a bit.
